File: ingestion/transform/entries.py
"""Load round_entries and session_entries."""

import pathlib
import sqlite3

from ingestion.transform import IngestState
from ingestion.transform.util import as_bool, as_float, as_int, read_csv


def load(con: sqlite3.Connection, dump_dir: pathlib.Path, state: IngestState) -> None:
    driver_map   = state.drivers   # jolpica_id → internal_id
    team_map     = state.teams
    race_number_by_round = state.race_number_by_round_jolpica_id  # jolpica round id → race_number

    # teamdriver: jolpica_id → {driver_id, team_id}
    td_map: dict[str, dict] = {}
    for r in read_csv(dump_dir, "formula_one_teamdriver.csv"):
        td_map[r["id"]] = {
            "driver_id": driver_map[int(r["driver_id"])],
            "team_id":   team_map[int(r["team_id"])],
        }

    # --- round_entries ---
    re_rows = []
    for r in read_csv(dump_dir, "formula_one_roundentry.csv"):
        race_number = race_number_by_round.get(r["round_id"])
        if race_number is None:
            continue  # cancelled round
        td = td_map.get(r["team_driver_id"])
        if td is None:
            continue  # defensive: should not happen
        re_rows.append((
            int(r["id"]),
            r["api_id"],
            race_number,
            td["driver_id"],
            td["team_id"],
            as_int(r, "car_number"),
        ))

    # All rows inserted — duplicates (shared drives / multi-car entries) are preserved
    # so their session_entries references remain valid. The race_results builder
    # deduplicates to the better result per (race_number, driver_id).
    con.executemany("""
        INSERT INTO round_entries
          (jolpica_id, jolpica_api_id, race_number, driver_id, team_id, car_number)
        VALUES (?,?,?,?,?,?)
    """, re_rows)

    # Build round_entry jolpica_id → internal_id map
    re_internal_map = {
        row["jolpica_id"]: row["id"]
        for row in con.execute("SELECT id, jolpica_id FROM round_entries")
    }

    # session jolpica_id → internal_id
    session_internal_map = {
        row["jolpica_id"]: row["id"]
        for row in con.execute("SELECT id, jolpica_id FROM sessions")
    }

    # --- session_entries ---
    se_rows = []
    for r in read_csv(dump_dir, "formula_one_sessionentry.csv"):
        session_id     = session_internal_map.get(int(r["session_id"]))
        round_entry_id = re_internal_map.get(int(r["round_entry_id"]))
        if session_id is None or round_entry_id is None:
            continue  # session or entry belongs to a cancelled round
        se_rows.append((
            int(r["id"]),
            r["api_id"],
            session_id,
            round_entry_id,
            as_int(r, "grid"),
            as_int(r, "position"),
            as_int(r, "laps_completed"),
            as_int(r, "status"),
            r.get("detail") or None,
            r.get("time") or None,
            as_float(r, "points"),
            as_bool(r, "is_classified"),
            as_bool(r, "is_eligible_for_points"),
            as_int(r, "fastest_lap_rank"),
        ))

    BATCH = 10_000
    for i in range(0, len(se_rows), BATCH):
        con.executemany("""
            INSERT INTO session_entries
              (jolpica_id, jolpica_api_id, session_id, round_entry_id,
               grid, position, laps_completed, status, detail, time,
               points, is_classified, is_eligible_for_points, fastest_lap_rank)
            VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?)
        """, se_rows[i:i + BATCH])

    # Expose for results/standings modules
    state.re_internal_map      = re_internal_map
    state.session_internal_map = session_internal_map

File: ingest/transform/entries.ts
import type { LibSQLDatabase } from 'drizzle-orm/libsql';
import { readCsv, asInt, asFloat, asBool } from '../csv';
import type { IngestState } from '../state';
import * as schema from '../../src/db/schema';

const CHUNK = 1000;

export async function loadEntries(
  db: LibSQLDatabase<typeof schema>,
  dumpDir: string,
  state: IngestState,
): Promise<void> {
  const { drivers: driverMap, teams: teamMap, raceNumberByRoundJolpicaId } = state;

  // Build teamdriver_id → {driver_id, team_id}
  const tdMap = new Map<string, { driverId: number; teamId: number }>();
  for (const r of readCsv(dumpDir, 'formula_one_teamdriver.csv')) {
    const driverId = driverMap.get(parseInt(r.driver_id, 10));
    const teamId = teamMap.get(parseInt(r.team_id, 10));
    if (driverId == null || teamId == null) continue;
    tdMap.set(r.id, { driverId, teamId });
  }

  // round_entries
  const reRows: typeof schema.roundEntries.$inferInsert[] = [];
  for (const r of readCsv(dumpDir, 'formula_one_roundentry.csv')) {
    const raceNumber = raceNumberByRoundJolpicaId.get(r.round_id);
    if (raceNumber == null) continue;  // cancelled round
    const td = tdMap.get(r.team_driver_id);
    if (!td) continue;
    // Keep ALL rows including shared-drive duplicates so session_entries FKs remain valid
    reRows.push({
      jolpicaId: parseInt(r.id, 10),
      jolpicaApiId: r.api_id,
      raceNumber,
      driverId: td.driverId,
      teamId: td.teamId,
      carNumber: asInt(r, 'car_number'),
    });
  }
  for (let i = 0; i < reRows.length; i += CHUNK) {
    await db.insert(schema.roundEntries).values(reRows.slice(i, i + CHUNK) as [typeof schema.roundEntries.$inferInsert, ...typeof schema.roundEntries.$inferInsert[]]);
  }

  // Build jolpica_id → internal_id maps
  const reInserted = await db
    .select({ id: schema.roundEntries.id, jolpicaId: schema.roundEntries.jolpicaId })
    .from(schema.roundEntries);
  state.reInternalMap = new Map(reInserted.map(r => [r.jolpicaId, r.id]));

  const sessInserted = await db
    .select({ id: schema.sessions.id, jolpicaId: schema.sessions.jolpicaId })
    .from(schema.sessions);
  state.sessionInternalMap = new Map(sessInserted.map(r => [r.jolpicaId, r.id]));

  // session_entries
  const seRows: typeof schema.sessionEntries.$inferInsert[] = [];
  for (const r of readCsv(dumpDir, 'formula_one_sessionentry.csv')) {
    const sessionId = state.sessionInternalMap.get(parseInt(r.session_id, 10));
    const roundEntryId = state.reInternalMap.get(parseInt(r.round_entry_id, 10));
    if (sessionId == null || roundEntryId == null) continue;  // cancelled round
    seRows.push({
      jolpicaId: parseInt(r.id, 10),
      jolpicaApiId: r.api_id,
      sessionId,
      roundEntryId,
      grid: asInt(r, 'grid'),
      position: asInt(r, 'position'),
      lapsCompleted: asInt(r, 'laps_completed'),
      status: asInt(r, 'status'),
      detail: r.detail || null,
      time: r.time || null,
      points: asFloat(r, 'points'),
      isClassified: asBool(r, 'is_classified'),
      isEligibleForPoints: asBool(r, 'is_eligible_for_points'),
      fastestLapRank: asInt(r, 'fastest_lap_rank'),
    });
  }
  for (let i = 0; i < seRows.length; i += CHUNK) {
    await db.insert(schema.sessionEntries).values(seRows.slice(i, i + CHUNK) as [typeof schema.sessionEntries.$inferInsert, ...typeof schema.sessionEntries.$inferInsert[]]);
  }
}
